ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -97,30 +97,32 @@ bech32Decode bech32 = do
     guard $ bech32VerifyChecksum hrp' dat'
     return (hrp', take (BS.length dat - 6) dat')
 
-toBase32 :: [Word8] -> [Word5]
-toBase32 dat = pad ++ result
+convertBits :: [Word] -> Int -> Int -> Bool -> Maybe [Word]
```

Comment:
There is a way to remove the use of `fromJust` but you might not like it.

```
type Pad f = Int -> [Word] -> [[Word]] -> f [[Word]]

yesPadding :: Pad Identity
yesPadding _ bits padValue result | bits /= 0 = return $ [padValue] : result
                                  | otherwise = return result
{-# INLINE yesPadding #-}

noPadding :: Pad Maybe
noPadding fromBits bits padValue result = do
  guard $ bits < fromBits && padValue == 0
  return result
{-# INLINE noPadding #-}

convertBits :: Functor f => [Word] -> Int -> Int -> Pad f -> f [Word]
{- ... -}
  go [] acc bits result = pad fromBits bits ((acc .<<. (tobits - bits)) .&. maxv) result
{- ... -}

toBase32 :: [Word8] -> [Word5]
toBase32 dat = map word5 $ runIdentity $ convertBits (map fromIntegral dat) 8 5 yesPadding
  
toBase256 :: [Word5] -> Maybe [Word8]
toBase256 dat = fmap (map fromIntegral) $ convertBits (map fromWord5 dat) 5 8 noPadding
```

If this technique excites you can use it, but I won't force it on you.  What you have already written is acceptable.

[#BooleanBlindness](https://shreevatsa.wordpress.com/2015/01/31/boolean-blindness/)